### PR TITLE
ATLAS-2933:Empty array attributes are returned as null instead of an empty list

### PR DESCRIPTION
**Pull request by mahaha000**

### CODE REVIEW

File: graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasJanusElement.java [hunk 1]
```diff
@@ -200,8 +204,8 @@ public String getIdForDisplay() {
 
         List<String> value = getListProperty(propertyName);
 
-        if (value.isEmpty()) {
-            return (List<V>)value;
+        if (value == null || value.isEmpty()) {
+            return value == null ? null : new ArrayList<>();
```
Comment (sarathsubramanian) [inline graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasJanusElement.java]: null check again here is not needed.

if (CollectionUtils.isEmpty(value)) {
  return value;
}

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java [hunk 2]
```diff
@@ -947,52 +947,63 @@ private AtlasEdge mapObjectIdValueUsingRelationship(AttributeMutationContext ctx
...
+            }
 
-                Map<String, Object> finalMap = removeUnusedMapEntries(attribute, ctx.getReferringVertex(), currentMap, newMap);
-                newMap.putAll(finalMap);
+            Map<String, Object> finalMap = removeUnusedMapEntries(attribute, ctx.getReferringVertex(), currentMap, newMap);
+            newMap.putAll(finalMap);
+        } else {
+            // primitive type map
+            if (isNewValNull) {
+                ctx.getReferringVertex().setProperty(propertyName, null);
```
Comment (sarathsubramanian) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java:996]: consider removing the property if new value is null - ctx.getReferringVertex().removeProperty(propertyName);

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java [hunk 3]
```diff
@@ -947,52 +947,63 @@ private AtlasEdge mapObjectIdValueUsingRelationship(AttributeMutationContext ctx
...
-
-                newVal.forEach((key, value) -> newMap.put(key.toString(), value));
             }
+            newVal.forEach((key, value) -> newMap.put(key.toString(), value));
+        }
 
-            if (isSoftReference) {
+        if (isSoftReference) {
+            if (isNewValNull) {
+                ctx.getReferringVertex().setProperty(propertyName,null);
```
Comment (sarathsubramanian) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java:1005]: => ctx.getReferringVertex().removeProperty(propertyName);

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java [hunk 4]
```diff
@@ -740,7 +741,9 @@ private Object mapVertexToAttribute(AtlasVertex entityVertex, AtlasAttribute att
         Map<String, AtlasObjectId> ret        = null;
         Map                        softRefVal = entityVertex.getProperty(attribute.getVertexPropertyName(), Map.class);
 
-        if (MapUtils.isNotEmpty(softRefVal)) {
+        if (softRefVal == null || MapUtils.isEmpty(softRefVal)) {
```
Comment (sarathsubramanian) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java]: MapUtils.isEmpty(softRefVal) checks for null and empty map, consider recoming softRefVal == null from if condition

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java [hunk 5]
```diff
@@ -751,15 +754,16 @@ private Object mapVertexToAttribute(AtlasVertex entityVertex, AtlasAttribute att
...
-
         return ret;
     }
 
     private List<AtlasObjectId> mapVertexToArrayForSoftRef(AtlasVertex entityVertex, AtlasAttribute attribute, AtlasEntityExtInfo entityExtInfo, final boolean isMinExtInfo) {
         List<AtlasObjectId> ret        = null;
         List                softRefVal = entityVertex.getListProperty(attribute.getVertexPropertyName(), List.class);
 
-        if (CollectionUtils.isNotEmpty(softRefVal)) {
+        if (softRefVal == null || CollectionUtils.isEmpty(softRefVal)) {
```
Comment (sarathsubramanian) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java]: same as above

**Comment by sarathkumarsubramanian** [comment]
Changes looks good. +1

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java (hunk 5 above)
Comment (simonpoortman) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphRetriever.java]: Simon Poortman

**Comment by nixonrodrigues** [comment]
This issue is fixed in ATLAS-2933, commit id https://gitbox.apache.org/repos/asf?p=atlas.git;a=commit;h=14037eed5c380006b0d622842677e362057a433a

Closing the PR

**Pull request closed without merging**